### PR TITLE
Allow optimizing over Bigfloat intervals

### LINKED ISSUE #43: Bug when using BigFloat interval
```jl
julia> minimise(x->(x^2 - 2)^2, big(-10..10))
ERROR: MethodError: no method matching filter_elements!(::HeapedVector{Tuple{Interval{BigFloat},Float64},IntervalOptimisation.var"#2#3"}, ::Tuple{Interval{BigFloat},BigFloat})
```

### PR DESCRIPTION
**Pull request by ericphanson**
I looked for something like the `numeric_type` function I introduce here in IntervalArithmetic.jl but didn't see anything, so I just defined it here. Fixes #43.

### CODE REVIEW

File: test/optimise.jl [hunk 1]
```diff
@@ -73,6 +85,12 @@ using Test
             @test global_max ⊆ 199.9..200
             m = (9.99..10)
             @test all(X ⊆ m × m || X ⊆ -m × m || X ⊆ m × -m || X ⊆ -m × -m for X in maximisers)
+
+            # same but with BigFloats
+            global_min, minimisers = minimise( X -> ( (x,y) = X; x^2 + y^2 ), (-big(10.0)..big(10.0)) × (-big(10.0)..big(10.0)), structure = Structure )
+            @test global_min ⊆ 0..1e-7
+            @test all(X ⊆ big(-1e-3..1e3) × big(-1e-3..1e-3) for X in minimisers)
```
Comment (ericphanson) [inline test/optimise.jl]: We need to make to use `big(-1e-3..1e3)` here due to the type constraint here:

https://github.com/JuliaIntervals/IntervalArithmetic.jl/blob/1516c5e1ca8e088aeafa9574072512aa2df241fb/src/multidim/intervalbox.jl#L69

Note that for the 1D tests, eg `@test global_min ⊆ -10 .. -9.999` works even when `global_min` is a Bigfloat interval. I think that type constraint could be lifted.

**Comment by dpsanders** [comment]
LGTM, thanks a lot!

Could you please bump the version in Project.toml to 0.4.1?
(I don't know if I can edit Pull Requests myself)

**Comment by dpsanders** [comment]
Apparently you can give me permission to modify your PR like this:
https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

But doesn't make sense for this particular one.

**Comment by ericphanson** [comment]
Sure, done. You won't be able to get it automerged though, since there's unbounded compat. Probably instead of
```julia
IntervalRootFinding = "≥ 0.4.0"
```
you want something like
```julia
IntervalRootFinding = "0.4, 0.5"
```
which means versions in the interval `[0.4, 0.6)`. I'm not really sure what versions are compatible though.

**Comment by dpsanders** [comment]
Hmm. Why does this package even need IntervalRootFinding?

Anyway, could you go ahead and make that change too please?

**Comment by ericphanson** [comment]
It doesn't, nor does it seem to need IntervalConstraintProgramming. I'll try and clean up the dependencies and fix the compat.

**Comment by ericphanson** [comment]
Turns out DataStructures, ForwardDiff, IntervalConstraintProgramming, and IntervalRootFinding were all unneeded. Makes the compat much easier to figure out :)

**Comment by dpsanders** [comment]
Most of those packages are needed by various of the open PRs. But then they should be added back in in those PRs I guess!

**Comment by dpsanders** [comment]
Thanks!

**Pull request merged**